AbstractSQLCompiler.ts

```diff
@@ -0,0 +1,347 @@
+import { AbstractSQLOptimiser } from './AbstractSQLOptimiser'
+import { AbstractSQLRules2SQL, Binding } from './AbstractSQLRules2SQL'
+const sbvrTypes = require('@resin/sbvr-types')
+import * as _ from 'lodash'
+import * as Promise from 'bluebird'
+
+interface AbstractSqlField {
+	fieldName: string
+	dataType: string
+	required: boolean
+	index: string
+	references?: {
+		tableName: string
+		fieldName: string
+	}
+	defaultValue?: string
+	necessity: boolean
+}
+interface AbstractSqlTable {
+	name: string
+	resourceName: string
+	idField: string
+	fields: AbstractSqlField[]
+	indexes: Array<{
+		type: string
+		fields: string[]
+	}>
+	primitive: false | string
+}
+interface ReferencedFields {
+	[alias: string]: string[]
+}
+interface SqlRule {
+	sql: string
+	bindings: Binding[]
+	structuredEnglish: string
+	referencedFields?: ReferencedFields
+}
+export interface AbstractSqlQuery extends Array<AbstractSqlQuery | string> {}
+export interface AbstractSqlModel {
+	synonyms: {
+		[ synonym: string ]: string
+	}
+	relationships: {}
+	tables: {
+		[ resourceName: string ]: AbstractSqlTable
+	}
+	rules: AbstractSqlQuery[]
+}
+interface SqlModel {
+	synonyms: {
+		[ synonym: string ]: string
+	}
+	relationships: {}
+	tables: {
+		[ resourceName: string ]: AbstractSqlTable
+	}
+	rules: SqlRule[]
+	createSchema: string[]
+	dropSchema: string[]
+}
+
+export enum Engines {
+	postgres = 'postgres',
+	mysql = 'mysql',
+	websql = 'websql',
+}
+
+const validateTypes = _.mapValues(sbvrTypes, ({ validate }) => {
+	if (validate != null) {
+		return Promise.promisify(validate)
+	}
+})
+
+const dataTypeValidate = (value: any, field: AbstractSqlField) => {
+	// In case one of the validation types throws an error.
+	const { dataType, required } = field
+	if (value == null) {
+		if (required) {
+			return Promise.reject('cannot be null')
+		} else {
+			return Promise.resolve(null)
+		}
+	} else if (validateTypes[dataType] != null) {
+		return validateTypes[dataType](value, required)
+	} else {
+		return Promise.reject('is an unsupported type: ' + dataType)
+	}
+}
+
+const dataTypeGen = (engine: Engines, { dataType, required, index, defaultValue }: AbstractSqlField): string => {
+	let requiredStr
+	if (required) {
+		requiredStr = ' NOT NULL'
+	} else {
+		requiredStr = ' NULL'
+	}
+	if (defaultValue != null) {
```

Comment:
Can we use _.isNull for all the null checks? It looks more consistent with the rest

Reply:
`_.isNull` specifically checks `=== null`, whereas we want `== null`, and for this specific case I prefer having the check directly as there are no special cases (whereas for a lot of types you have a bunch of hoops to jump through to check)